agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h

```diff
@@ -242,21 +244,33 @@ class backhaul_manager : public btl::transport_socket_thread {
 
     std::unique_ptr<beerocks::agent_ucc_listener> m_agent_ucc_listener;
 
+    struct sClientInfo {
+        sMacAddr client_mac;
+        sMacAddr radio_mac; // ruid
+        sMacAddr bssid;     // VAP mac
+        std::chrono::steady_clock::time_point time_stamp;
+        size_t asso_len;
+        uint8_t assoc_req[ASSOCIATION_FRAME_SIZE];
```

Comment:
```suggestion
        std::array<uint8_t> assoc_req;
```

If you see a size and array combination in any structure that is not tlv-generated, it's probably wrong :-). Actually, if you see a C-style array it's probably wrong.

Hm, maybe I need to make a CONTRIBUTING.md update...

Note that with the std::array, the copy is simply `std::copy(s.assoc_req, client_capability_report_tlv->association_frame())`.

Reply:
I agree it does not make sense to use size and array combination inside a structure.
The reason I did it is that I need to know the length of each association from so when I pass it in the client capability report it will have the correct length.

One option I thought about is to use a vector, which is dynamic but it isn't very comfortable to use (to push every time another `uint8_t`).
Another is to declare it as c array and to fill it with a null terminator (as we did in the ap_wlan_hal Event::STA_Connected) and then we can use `strnlen` after converting to string.

but `std::array` also requires a length argument so it's also not a dynamic array.
I want to know the length of each association frame and not take the fixed size, so how does `std::array` help me here? let me know if I did not understand your suggestion @arnout

Reply:
OK, I was mixing things up a little, sorry about that...

`uint8_t assoc_req[ASSOCIATION_FRAME_SIZE];` is equivalent to `std::array<uint8_t, ASSOCIATION_FRAME_SIZE> assoc_req;` - but the latter would still require asso_len. So in this case, it should indeed be a `std::vector`.

However, as you observe, a vector is not very comfortable, because we later want to copy it out with set_association_frame(). Well, actually, it's not that bad, because we can use `std::copy_n` to fill in the vector, and `std::vector::data()` to access the data directly for set_association_frame().

However, as I indicated before, I think it's better to use a `basic_string<uint8_t>` for such a situation.

I'm working on a quick commit to show what I mean.

Reply:
I pushed my proposal to dev/client-caps-report-assoc_req-with-bytes (additional fixup commits on top of this PR). If it looks good to @CoralMalachi and @tomereli just take those commits in this PR and rebase on master. I didn't create a PR for that branch.

Note that this is not terribly important because we already use this C-style array pattern in dozens of places in the code, so adding one more is not a huge concern. And we really should get this PR merged soonish, it's been dragging on for too long..

Reply:
I like it. I basic_string is a weird name for storing binary data so using a bytes class is welcome.